### PR TITLE
Raise event from aggregate root

### PR DESCRIPTION
**Pull request by mbadolato**
```
Entity objects must either extend AggregateRoot or DomainEventProvider.
Since the usage of Doctrine and/or CRUD plugins requires the use of
AggregateRoot, there is no ability to call raise() to execute a
DomainObjectChanged notice. which means that we cannot have event
handlers listening for DomainObjectChanged.

Using apply() does not allow us to send DomainObjectChanged objects
since it would subsequently then require an applyDomainObjectChanged
method, which is unneeded.

apply() and raise() do the exact same thing other than the apply()
method calling $this->executeEvent($event), thus it's a simple matter
to add raise() to AggregateRoot, which then enables us to do

$this->raise(new DomainObjectChanged("ChangeEmail", array("email" => $email, "oldEmail" => $this->email)));

from our Entity objects which extend AggregateRoot so that they can
use the Doctrine and CRUD plugins.
```

### CODE REVIEW

**Comment by mbadolato** [comment]
@beberlei Can you merge in these small fixes/changes so that I don't have to track my local fork? Thx

File: src/LiteCQRS/AggregateRoot.php [hunk 1]
```diff
@@ -31,6 +31,11 @@ public function dequeueAppliedEvents()
     protected function apply(DomainEvent $event)
     {
         $this->executeEvent($event);
+        $this->raise($event);
+    }
+
+    protected function raise(DomainEvent $event)
```
Comment (beberlei) [inline src/LiteCQRS/AggregateRoot.php:37]: i dont like this one, that is what DomainEventProvider is for, the AggregateRoot is specficially for apply() not for raise()

Comment (mbadolato) [inline src/LiteCQRS/AggregateRoot.php:37]: So if I need to extend AggregateRoot and I want an event handler to listen for a DomainObjectChange and do something, how?

As far as I have been able to tell, there isn't a way to get apply() to trigger the events into the onFoo event class

Comment (beberlei) [inline src/LiteCQRS/AggregateRoot.php:37]: You don't use the AggregateRoot for this case, AggregateRoot is specifically for event sourcing, where you need to apply all events again. If you don't use event sourcing, you can use https://github.com/beberlei/litecqrs-php/blob/master/src/LiteCQRS/DomainEventProvider.php instead of AggregateRoot.

Comment (mbadolato) [inline src/LiteCQRS/AggregateRoot.php:37]: But using $this->repository->add($object) (if using the Doctrine ORM plugin) doesn't work if the entity class extends DomainEventProvider; it requires AggregateRootInterface :)

Thus you cannot use raise() and apply() with the same Entity

Comment (beberlei) [inline src/LiteCQRS/AggregateRoot.php:37]: Yes, but that is the problem, they need `EventProviderInterface` as a typehint then instead of breaking the API here.

Comment (mbadolato) [inline src/LiteCQRS/AggregateRoot.php:37]: So, just to be clear, you're saying  I should submit a different PR changing `ORMRepository::add(...)` and `update(...)` to type hint with `EventProviderInterface` instead of `AggregateRootInterface` then?

Comment (beberlei) [inline src/LiteCQRS/AggregateRoot.php:37]: yes, please create a `DomainEventProviderRepository` interface, that looks like the `AggregateRootRepositoryInterface` and mark the later one as `@deprecated`. And then change all the implementations to use `DomainEventProviderRepository` with the typehint for `EventProviderInterface` instead.

Comment (mbadolato) [inline src/LiteCQRS/AggregateRoot.php:37]: Ok, I will look at doing this tomorrow evening. Thanks.

Comment (mbadolato) [inline src/LiteCQRS/AggregateRoot.php:37]: I had a little more time just now. PR-41 submitted.
